tree/ntuple/v7/inc/ROOT/RNTupleProcessor.hxx

```diff
@@ -0,0 +1,194 @@
+      std::unique_ptr<ROOT::Experimental::RFieldBase> fConcreteField;
+      REntry::RFieldToken fToken;
+      std::shared_ptr<void> fValuePtr;
+
+      RFieldContext(std::unique_ptr<ROOT::Experimental::RFieldBase> protoField, REntry::RFieldToken token)
+         : fProtoField(std::move(protoField)), fToken(token)
+      {
+      }
+
+      void ResetConcreteField() { fConcreteField = nullptr; }
+      void SetConcreteField() { fConcreteField = fProtoField->Clone(fProtoField->GetFieldName()); }
+   };
+
+   std::unique_ptr<RNTupleModel> fModel;
+   std::unique_ptr<REntry> fEntry; ///< The entry is based on the first page source
+   std::unique_ptr<Internal::RPageSource> fPageSource;
+   std::vector<RFieldContext> fFieldContexts;
+
+public:
+   RNTupleProcessorContext(const RNTupleSourceSpec &ntuple);
+
+   /////////////////////////////////////////////////////////////////////////////
+   /// \brief Initializes the processor context
+   ///
+   /// Recursively creates so-called 'protofields' based on the model of the first page source, which will serve as
+   /// a basis for all subsequent RNTuples to be processed.
+   void Initialize();
+
+   /////////////////////////////////////////////////////////////////////////////
+   /// \brief Updates the processor context
+   ///
+   /// \param[in] ntuple The RNTuple to connect for processing
+   ///
+   /// \return The number of entries in the newly-connected RNTuple
+   ///
+   /// Creates and attaches new page source for the specified RNTuple, and connects the fields that are known by
+   /// the processor context to it.
+   NTupleSize_t Update(const RNTupleSourceSpec &ntuple);
+
+   /////////////////////////////////////////////////////////////////////////////
+   /// \brief Creates and connects concrete fields to the current page source, based on the protofields
+   void ConnectFields();
+
+   /////////////////////////////////////////////////////////////////////////////
+   /// \brief Loads an entry
+   ///
+   /// \param[in] index The index (local to the current RNTuple) of the entry to load.
+   void LoadEntry(NTupleSize_t index) { fEntry->Read(index); }
+
+   /////////////////////////////////////////////////////////////////////////////
+   /// \brief Returns a reference to the entry in this context
+   ///
+   /// \return A reference to the entry from this context
+   ///
+   const REntry &GetEntry() const { return *fEntry; }
+
+   /////////////////////////////////////////////////////////////////////////////
+   /// \brief Returns a reference to the page source currently used in this context
+   ///
+   /// \return A reference to the page source
+   Internal::RPageSource &GetPageSource() const { return *fPageSource; }
+};
+} // namespace Internal
+
+class RNTupleProcessor {
+private:
+   const std::vector<RNTupleSourceSpec> &fNtuples;
+   std::unique_ptr<REntry> fEntry; ///< The entry is based on the first page source
+   std::unique_ptr<Internal::RNTupleProcessorContext> fContext;
```

Comment:
I'm not quite sure why we need the `RNTupleProcessorContext`. Couldn't its functionality just be part of the `RNTupleProcessor` itself?